math/src/field/f23201/mod.rs

```diff
@@ -0,0 +1,512 @@
+    slice,
+};
+use utils::{
+    collections::Vec, string::ToString, AsBytes, ByteReader, ByteWriter, Deserializable,
+    DeserializationError, Randomizable, Serializable,
+};
+
+#[cfg(test)]
+mod tests;
+
+// CONSTANTS
+// ================================================================================================
+
+/// Field modulus = 2^23 - 2^20 + 1
+const M: u32 = 0x700001;
+
+/// R^2 mod M; this is used for conversion of elements into Montgomery representation.
+const R2: u32 = 0x32f054;
+
+/// 2^20 primitive root of unity
+const G: u32 = 2187;
+
+/// Number of bytes needed to represent field element
+const ELEMENT_BYTES: usize = 3;
+
+// FIELD ELEMENT
+// ================================================================================================
+
+/// Represents base field element in the field.
+///
+/// Internal values are stored in the range [0, 2M). The backing type is `u32`.
+#[derive(Copy, Clone, Debug, Default)]
+pub struct BaseElement(u32);
+impl BaseElement {
+    /// Creates a new field element from the provided `value`; the value is converted into
+    /// Montgomery representation.
+    pub const fn new(value: u32) -> BaseElement {
+        Self(mont_red_cst((value as u64) * (R2 as u64)))
+    }
+
+    /// Returns a new field element from the provided 'value'. Assumes that 'value' is already
+    /// in canonical Montgomery form.
+    pub const fn from_mont(value: u32) -> BaseElement {
+        BaseElement(value)
+    }
+
+    /// Returns the non-canonical u32 inner value.
+    pub const fn inner(&self) -> u32 {
+        self.0
+    }
+
+    /// Computes an exponentiation to the power 7. This is useful for computing Rescue-Prime
+    /// S-Box over this field.
+    #[inline(always)]
+    pub fn exp7(self) -> Self {
+        let x2 = self.square();
+        let x4 = x2.square();
+        let x3 = x2 * self;
+        x3 * x4
+    }
+}
+
+impl FieldElement for BaseElement {
+    type PositiveInteger = u64; // TODO can we convert this to u32?
+    type BaseField = Self;
+
+    const ZERO: Self = Self::new(0);
+    const ONE: Self = Self::new(1);
+
+    const ELEMENT_BYTES: usize = ELEMENT_BYTES;
+    const IS_CANONICAL: bool = false;
+
+    #[inline]
+    fn double(self) -> Self {
+        Self(reduce_le_2m(self.0 + self.0))
+    }
+
+    #[inline]
+    fn exp(self, power: Self::PositiveInteger) -> Self {
+        let mut b: Self;
+        let mut r = Self::ONE;
+        for i in (0..64).rev() {
+            r = r.square();
+            b = r;
+            b *= self;
+            // Constant-time branching
+            let mask = -(((power >> i) & 1 == 1) as i32) as u32;
+            r.0 ^= mask & (r.0 ^ b.0);
+        }
+
+        r
+    }
+
+    #[inline]
+    #[allow(clippy::many_single_char_names)]
+    fn inv(self) -> Self {
+        // M-2 = 0x6fffff
+        let mut out = (self.square() * self).square();
+        for _ in 0..20 {
+            out = out.square() * self;
+        }
+        out
```

Comment:
Could you add this TODO as a comment?

```
 $ addchain search 7340031
addchain: expr: "7340031"
addchain: hex: 6fffff
addchain: dec: 7340031
addchain: best: opt(dictionary(sliding_window(32),continued_fractions(dichotomic)))
_10    = 2*1
_11    = 1 + _10
_110   = 2*_11
_11000 = _110 << 2
_11110 = _110 + _11000
x9     = _11110 << 4 + _11110 + 1
i15    = 2*x9 + x9 + _11
x11    = x9 + i15
return   (i15 + x11) << 11 + x11
```

Reply:
https://github.com/bwesterb/winterfell/pull/2